### PR TITLE
340 interface additional cf returns

### LINKED ISSUE #148: Refactoring process_metadata
Right now the following code-block gets copied for every CR that has an `out` datatype which has a continuous time-axis:
```python
        # Some index gymnastics to get trial begin/end "samples"                      
        if data._selection is not None:                                               
            chanSec = data._selection.channel                                         
            trl = data._selection.trialdefinition                                     
            for row in range(trl.shape[0]):                                           
                trl[row, :2] = [row, row + 1]                                         
        else:                                                                         
            chanSec = slice(None)                                                     
            time = np.arange(len(data.trials))                                        
            time = time.reshape((time.size, 1))                                       
            trl = np.hstack((time, time + 1,                                          
                             np.zeros((len(data.trials), 1)),                         
                             np.array(data.trialinfo)))                               
                                                                                      
        # Attach constructed trialdef-array (if even necessary)                       
        if self.keeptrials:                                                           
            out.trialdefinition = trl                                                 
        else:                                                                         
            out.trialdefinition = np.array([[0, 1, 0]])
```
Would be nice to bind that somewhere higher up in the class hierarchy or at least write an auxiliary function.

### LINKED ISSUE #340: Interface additional cF returns
Now that we finally have the possibility to get more return values from our cFs thanks to #321 we should make use of it:

### Granger

- the Wilson factorization returns a boolean `conv` indicating successful termination of the algorithm which might be of interest
- Potential regularization of the CSD also returns the `factor`

### Specest Methods

- everywhere where single trial `freq` axis appear, we could/should at least compare the hashes as in the end there is one global `.freq`

### PR DESCRIPTION
**Pull request by tensionhead**
## Make use of additional cF returns

- Granger now returns 4 additional metadata entries stored in `.info` documenting the factorization
- equality of frequency axis for mtmfft and csd get checked via hashes
- Further refactored `process_metadata` to reduce copied code and have one place where the properties get attached from input to output (`propagate_properties`)

closes #148
closes #340 

Author Guidelines
-----------------
- [x] Is the change set **< 600 lines**?
- [x] Was the code checked for memory leaks/performance bottlenecks?
- [x] Is the code running locally **and** on the ESI cluster?
- [x] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [x] Are testing routines present?
- [x] Do **parallel** loops have a set length and correct termination conditions?
- [x] Do objects in the global package namespace perform proper parsing of their input? 
- [x] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [x] Code layout
  - [x] Is the code PEP8 compliant?
  - [x] Does the code adhere to agreed-upon naming conventions?
  - [x] Are keywords clearly named and easy to understand?
  - [x] No commented-out code?
- [x] Are all docstrings complete and accurate?
- [x] Is the CHANGELOG.md up to date?

### CODE REVIEW

File: syncopy/specest/compRoutines.py [hunk 1]
```diff
@@ -200,31 +206,17 @@ class MultiTaperFFT(ComputationalRoutine):
...
-            trl = np.hstack((time, time + 1,
-                             np.zeros((len(data.trials), 1)),
-                             np.array(data.trialinfo)))
-
-        # Attach constructed trialdef-array (if even necessary)
-        if self.keeptrials:
-            out.trialdefinition = trl
-        else:
-            out.trialdefinition = np.array([[0, 1, 0]])
+        check_freq_hashes(metadata, out)
```
Comment (dfsp-spirit) [inline syncopy/specest/compRoutines.py:211]: good idea to put this into a function

File: syncopy/shared/computational_routine.py [hunk 2]
```diff
@@ -1019,15 +1019,14 @@ def process_metadata(self, data, out):
 
 # --- metadata helper functions ---
 
-def propagate_metadata(in_data, out_data):
+def propagate_properties(in_data, out_data, keeptrials=True):
```
Comment (dfsp-spirit) [inline syncopy/shared/computational_routine.py:1022]: very nice, this reduces quite a bit code duplication

File: syncopy/tests/backend/test_conn.py [hunk 3]
```diff
@@ -236,15 +234,13 @@ def test_regularization():
...
-    assert CN > cmax
-
     # --- regularize CSD ---
 
-    CSDreg, fac = regularize_csd(CSD, cond_max=cmax, nSteps=25)
+    cmax = 1e4
+    CSDreg, fac, iniCN = regularize_csd(CSD, cond_max=cmax, nSteps=25)
+    # get initial CSD condition number, which is way too large!
+    print(iniCN)
+    assert iniCN > cmax
```
Comment (dfsp-spirit) [inline syncopy/tests/backend/test_conn.py:243]: would be nice to use 2nd arg of assert statement to print better error here, like

   assert iniCN > cmax, f"some message about var {var} here"

**Comment by dfsp-spirit** [comment]
Looks good,

only 2 things:

* the changelog should be updated
* do you want to document the new Granger return values in the docstrings of the frontend, so that the users know about them?

File: syncopy/shared/computational_routine.py (hunk 2 above)
Comment (tensionhead) [reply syncopy/shared/computational_routine.py:1022]: As discussed, it's still not perfect imho.. I could also think of putting those property mappings into the respective module `compRoutines.py`, however there is already quite some crossover between `freqanalysis` and `connectivityanalysis` for example.

File: syncopy/tests/backend/test_conn.py (hunk 3 above)
Comment (tensionhead) [reply syncopy/tests/backend/test_conn.py:243]: yes, I've seen this only in your code now.. did not know this is possible, but good idea :+1:

Comment (tensionhead) [reply syncopy/tests/backend/test_conn.py:243]: puh, there was definitely something off in any case.. I corrected it, still not sure how often ill conditioned CSDs appear in the wild :sweat_smile: 
But at least now we know at least _aposteriori_ what the factorization/regularization did thanks to the additional cF returns :)

**Comment by tensionhead** [comment]
About your 2nd point regarding the docstring of the frontend, I now think we should programmatically fix/define the key's which are propagated from the CR/cF to `.info`. Then also Sphinx can pick them up easily via `:data:`syncopy.nwanalysis.....`, but for now I 'hardcoded' the `.info` keys into the frontend doc.

**Comment by tensionhead** [comment]
I also added the changelog entries for the additional fooof returns, which are also new right :) ?! Feel free to edit/change!

**Pull request merged**